File: hazelcastcsharpclientproblem/dotnetclienttopiclistener/HazelcastDotNetProblem/MyPortableFactory.cs
using Hazelcast.IO.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace HazelcastDotNetProblem
{
    public class MyPortableFactory : IPortableFactory
    {
        public IPortable Create(int classId)
        {
            if (classId == 1)
            {
                return new TopicData();
            }
            else
            {
                throw new NotImplementedException(classId + " unsupported classId");
            }

        }

        public MyPortableFactory()
        {

        }

        public static int FactoryId = 1;
    }
}

File: hazelcastcsharpclientproblem/clusterandjavaproducer/hazelcastproblemdotnet/common/src/main/java/com/mycompany/common/MyPortableFactory.java
package com.mycompany.common;

import com.hazelcast.nio.serialization.Portable;
import com.hazelcast.nio.serialization.PortableFactory;

public class MyPortableFactory implements PortableFactory {

    /**
     * Create an instance of a Portable matching the provided classId.
     */
    @Override
    public Portable create(int classId) {
        if (classId == 1) {
            return new TopicData();
        }  else {
            throw new IllegalArgumentException(classId + " unsupported classId");
        }

    }

}
